src/Microsoft.Tye.Core/ApplicationFactory.cs

```diff
@@ -149,7 +149,15 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
                     }
                     else if (service is ContainerServiceBuilder container)
                     {
-                        container.EnvironmentVariables.Add(envVar);
+                        if (configEnvVar.Value is null)
+                        {
+                            var value = UserSecretsBuilder.GetSecretValue(configEnvVar.SecretStoreId, configEnvVar.SecretName);
```

Comment:
I think this should be resolved later, preserve the secret ness until the runner or deployment model needs it. @rynowak ideas here?

Reply:
Yea I like this idea. There should be a secret resolver component that runs later and does the right thing. It shouldn’t happen at the configuration layer

Reply:
I feel like I want to add a new file to manage secrets and secret provider info so `tye.yaml` doesn't blow up. This would leave the config model untouched. Since the connection settings for something like a key vault provider might contain some sensitive information itself having those settings in a separate file would also allow that file to be easily excluded from source control.
 
In the hosting layer, right before a service is started, the secret resolver would check for any secrets and add new env vars. If the data is in tye.yaml, then that file should either be re-read to get secret info (so the config model is not dealing with any secret stuff) or some info about the secrets (but not the actual secret value) should be added to the config model and passed through up to the hosting layer so that the value can be resolved before running the service.

Pros for a secret file:
- keep secrets totally separate from config
- might also be useful to help tooling generate k8 manifests
- keep sensitive info out of SCM easily
- secret management command(s) you mention like `tye secret add ...` could work in this file
- [package references](https://github.com/dotnet/tye/issues/29) could strip out sensitive info and pack the secrets file so that consumers could just fill in the sensitive info and get up and running.

Cons for a secret file:
- extra file to parse and manage schema for
-  invites the possibility of having an env in `tye.yaml` and a secret with the same env var name so would require some kind of conflict resolution
- I'm sure there are others

For example:
```
name: microservice
secretProviders:
  - name: user-secrets
    type: user-secrets
    settings:
      secretId: ABC
  - name: other-user-secrets
    type: user-secrets
    settings:
      secretId: XYZ
  - name: key-vault
    type: key-vault
    settings:
      vaultName: MyAzureKeyVault
services:
  - name: redis    
    env:    
      - name: SECRET_1
        secretProvider: user-secrets
        secretKey: REDIS_SECRET
      - name: SECRET_2
        secretProvider: other-user-secrets
        secretKey: REDIS_SECRET
      - name: SECRET_3
        secretProvider: key-vault
        secretKey: REDIS_SECRET
```

would result in 3 env vars being added to the redis container
```
SECRET_1=<value of REDIS_SECRET from user secrets ABC>
SECRET_2=<value of REDIS_SECRET from user secrets XYZ>
SECRET_3=<value of REDIS_SECRET from key vault MyAzureKeyVault>
```